### PR TITLE
cluster: add definition and lock config types

### PR DESCRIPTION
**Pull request by corverroos**
Adds the `cluster` package and the two config types: `Definition` and `Lock`. Also do basic json serialisation and hash based validation. Also implement EIP712 signatures and verification.

category: feature
ticket: none
feature_set: alpha

### CODE REVIEW

File: cluster/cluster.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	"io"
+)
+
+const (
+	paramsVersion = "v1.0.0"
+	dkgAlgo       = "default"
+)
+
+// NewParams returns a new params with populated version and UUID.
+func NewParams(name string, numVals int, threshold int,
```
Comment (leolara) [inline cluster/cluster.go:39]: @corverroos parameters to a function should be either all in one line or each in its own line. Doesn't linter detect this?

File: cluster/types.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cluster
+
+// Params defines an intended charon cluster configuration.
+type Params struct {
```
Comment (leolara) [inline cluster/types.go:19]: @corverroos the public fields make this structure mutable, unfortunatelly, making it less functional. But I understand that to do the json encoding would require a bit of boilerplate code if we made them private, so perhaps it is acceptable?

File: cluster/types.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
+// Copyright © 2022 Obol Labs Inc.
```
Comment (leolara) [inline cluster/types.go:1]: @corverroos it is cleaner to have a params.go operator.go lock.go  and distvalidator.go and then the methods of each in that file

File: cluster/encode.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+import (
+	"bytes"
+	"encoding/json"
+	"strings"
+
+	"github.com/obolnetwork/charon/app/errors"
+)
+
+func (p Params) MarshalJSON() ([]byte, error) {
+	type fieldsOnly Params // Marshal fields-only version of params
```
Comment (leolara) [inline cluster/encode.go:27]: @corverroos what does this renaming of Params?

File: cluster/types.go (hunk 2 above)
Comment (leolara) [reply cluster/types.go:19]: considering we are already writing a MarshallJSON and UnmarshallJSOn for Params, there is no need for these fields to be public?

**Review by leolara — CHANGES_REQUESTED** [review]

File: cluster/types.go (hunk 2 above)
Comment (corverroos) [reply cluster/types.go]: So this is a DTO, ie. a pure value, fields should be public.

File: cluster/types.go (hunk 3 above)
Comment (corverroos) [reply cluster/types.go:1]: I chose the option to colocate logic, almost like how eth2client does things.

File: cluster/encode.go (hunk 4 above)
Comment (corverroos) [reply cluster/encode.go]: I've refactored this to use json formatter type, same as `Lock`

File: cluster/cluster.go (hunk 1 above)
Comment (corverroos) [reply cluster/cluster.go]: sure, will make it flat.

**Comment by leolara** [comment]
I don't think we should add everything as a new package in the root. I am not sure about this one, but perhaps we need to think about a package hierarchy and where to put certain packages that do not fit on app or core.

**Comment by corverroos** [comment]
It is just 4 types, I do not think creating more packages is required.

File: cluster/encode.go (hunk 4 above)
Comment (leolara) [reply cluster/encode.go]: ok, I didn't understood this version

File: cluster/types.go (hunk 3 above)
Comment (leolara) [reply cluster/types.go:1]: in eth2client they subsecuently split things of a component into subjects (as they do in prysm) for the same component, but I don't see they do exactly what you do here of putting every component together and then put the same subject of all the components in the same file.

**Comment by leolara** [comment]
I am not talking about creating more packages, folders can be empty, my argument is about not putting all new packages in the root.

File: cluster/types.go (hunk 2 above)
Comment (corverroos) [reply cluster/types.go]: ok, grouping functions by type.

File: cluster/types.go (hunk 3 above)
Comment (corverroos) [reply cluster/types.go]: ok, grouping files by type. So now a single file for each type

**Comment by corverroos** [comment]
`cluster` package is in root because it is a top level thing. The cluster configuration. It will be used by `dkg` and `app` packages. So I do not think it should be in `app/cluster`, neither `cmd/cluster` or `dkg/cluster`.

File: cluster/types.go (hunk 3 above)
Comment (leolara) [reply cluster/types.go]: Thanks

**Comment by leolara** [comment]
Approved

**Pull request merged**